File: app/src/main/java/com/elishaazaria/sayboard/settingsfragments/modelsfragment/ModelsAdapter.java
package com.elishaazaria.sayboard.settingsfragments.modelsfragment;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import com.elishaazaria.sayboard.R;
import com.elishaazaria.sayboard.downloader.messages.ModelInfo;

import java.util.ArrayList;
import java.util.List;

public class ModelsAdapter extends RecyclerView.Adapter<ModelsAdapter.ViewHolder> {

    public enum DataState {
        CLOUD, INSTALLED, DOWNLOADING, QUEUED
    }

    private final Context context;
    private final AdapterDataProvider dataProvider;
    private final List<ModelsAdapterData> mData;
    private final LayoutInflater mInflater;
    private ItemClickListener mClickListener;

    // data is passed into the constructor
    public ModelsAdapter(Context context, AdapterDataProvider dataProvider) {
        this.mInflater = LayoutInflater.from(context);
        this.context = context;
        this.dataProvider = dataProvider;
        mData = new ArrayList<>(dataProvider.getData());
    }

    // inflates the row layout from xml when needed
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = mInflater.inflate(R.layout.fragment_models_entry, parent, false);
        return new ViewHolder(view);
    }

    // binds the data to the TextView in each row
    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        ModelsAdapterData data = mData.get(position);
        holder.titleTextView.setText(data.getTitle());
        holder.subtitleTextView.setText(data.getSubtitle());
        holder.downloadButton.setImageResource(data.getImageRes());

        holder.data = data;
    }

    // total number of rows
    @Override
    public int getItemCount() {
        return mData.size();
    }


    // stores and recycles views as they are scrolled off screen
    public class ViewHolder extends RecyclerView.ViewHolder {
        TextView titleTextView;
        TextView subtitleTextView;
        ImageButton downloadButton;
        ModelsAdapterData data;

        ViewHolder(View itemView) {
            super(itemView);
            titleTextView = itemView.findViewById(R.id.titleTextView);
            subtitleTextView = itemView.findViewById(R.id.subtitleTextView);
            downloadButton = itemView.findViewById(R.id.downloadButton);

            downloadButton.setOnClickListener(this::onButtonClick);
            itemView.setOnClickListener(this::onClick);
        }

        private void onClick(View view) {
            if (mClickListener != null)
                mClickListener.onItemClick(view, getAdapterPosition(), data);
        }

        private void onButtonClick(View view) {
            if (mClickListener != null) {
                mClickListener.onButtonClicked(view, getAdapterPosition(), data);
            }
        }
    }

    // convenience method for getting data at click position
    public ModelsAdapterData getItem(int index) {
        return mData.get(index);
    }

    public int size() {
        return mData.size();
    }

    public boolean changed(ModelsAdapterData data) {
        int index = mData.indexOf(data);
        if (index == -1) return false;
        notifyItemChanged(index);
        return true;
    }

    public boolean removed(ModelsAdapterData data) {
        int index = mData.indexOf(data);
        if (index == -1) return false;
        mData.remove(index);
        notifyItemRemoved(index);
        return true;
    }

    @SuppressLint("NotifyDataSetChanged")
    public void reload() {
        mData.clear();
        mData.addAll(dataProvider.getData());
        notifyDataSetChanged();
    }

    // allows clicks events to be caught
    public void setClickListener(ItemClickListener itemClickListener) {
        this.mClickListener = itemClickListener;
    }

    // parent activity will implement this method to respond to click events
    public interface ItemClickListener {
        void onItemClick(View view, int position, ModelsAdapterData data);

        void onButtonClicked(View view, int position, ModelsAdapterData data);
    }
}

File: app/src/main/java/com/elishaazaria/sayboard/settingsfragments/modelsfragment/ModelsAdapter.kt
package com.elishaazaria.sayboard.settingsfragments.modelsfragment

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.elishaazaria.sayboard.R

class ModelsAdapter(context: Context, dataProvider: AdapterDataProvider) :
    RecyclerView.Adapter<ModelsAdapter.ViewHolder>() {
    enum class DataState {
        CLOUD, INSTALLED, DOWNLOADING, QUEUED
    }

    private val context: Context
    private val dataProvider: AdapterDataProvider
    private val mData: MutableList<ModelsAdapterData>
    private val mInflater: LayoutInflater
    private var mClickListener: ItemClickListener? = null

    // data is passed into the constructor
    init {
        mInflater = LayoutInflater.from(context)
        this.context = context
        this.dataProvider = dataProvider
        mData = ArrayList(dataProvider.data)
    }

    // inflates the row layout from xml when needed
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = mInflater.inflate(R.layout.fragment_models_entry, parent, false)
        return ViewHolder(view)
    }

    // binds the data to the TextView in each row
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val data = mData[position]
        holder.titleTextView.text = data.title
        holder.subtitleTextView.text = data.subtitle
        holder.downloadButton.setImageResource(data.imageRes)
        holder.data = data
    }

    // total number of rows
    override fun getItemCount(): Int {
        return mData.size
    }

    // stores and recycles views as they are scrolled off screen
    inner class ViewHolder internal constructor(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        var titleTextView: TextView
        var subtitleTextView: TextView
        var downloadButton: ImageButton
        var data: ModelsAdapterData? = null

        init {
            titleTextView = itemView.findViewById(R.id.titleTextView)
            subtitleTextView = itemView.findViewById(R.id.subtitleTextView)
            downloadButton = itemView.findViewById(R.id.downloadButton)
            downloadButton.setOnClickListener { view: View -> onButtonClick(view) }
            itemView.setOnClickListener { view: View -> onClick(view) }
        }

        private fun onClick(view: View) {
            if (mClickListener != null) mClickListener!!.onItemClick(view, adapterPosition, data!!)
        }

        private fun onButtonClick(view: View) {
            if (mClickListener != null) {
                mClickListener!!.onButtonClicked(view, adapterPosition, data!!)
            }
        }
    }

    // convenience method for getting data at click position
    fun getItem(index: Int): ModelsAdapterData {
        return mData[index]
    }

    fun size(): Int {
        return mData.size
    }

    fun changed(data: ModelsAdapterData?): Boolean {
        val index = mData.indexOf(data)
        if (index == -1) return false
        notifyItemChanged(index)
        return true
    }

    fun removed(data: ModelsAdapterData?): Boolean {
        val index = mData.indexOf(data)
        if (index == -1) return false
        mData.removeAt(index)
        notifyItemRemoved(index)
        return true
    }

    @SuppressLint("NotifyDataSetChanged")
    fun reload() {
        mData.clear()
        mData.addAll(dataProvider.data)
        notifyDataSetChanged()
    }

    // allows clicks events to be caught
    fun setClickListener(itemClickListener: ItemClickListener?) {
        mClickListener = itemClickListener
    }

    // parent activity will implement this method to respond to click events
    interface ItemClickListener {
        fun onItemClick(view: View, position: Int, data: ModelsAdapterData)
        fun onButtonClicked(view: View, position: Int, data: ModelsAdapterData)
    }
}